localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -1797,11 +1842,96 @@ def create_event_source_mapping(
         context: RequestContext,
         request: CreateEventSourceMappingRequest,
     ) -> EventSourceMappingConfiguration:
-        service = None
+        if config.LAMBDA_EVENT_SOURCE_MAPPING == "v2":
+            event_source_configuration = self.create_event_source_mapping_v2(context, request)
+        else:
+            event_source_configuration = self.create_event_source_mapping_v1(context, request)
 
+        return event_source_configuration
+
+    def create_event_source_mapping_v2(
+        self, context: RequestContext, request: CreateEventSourceMappingRequest
+    ) -> EventSourceMappingConfiguration:
+        # Validations
+        function_arn, function_name, state = self.validate_event_source_mapping(context, request)
+
+        esm_config = EsmConfigFactory(request, function_arn).get_esm_config()
+
+        # Copy esm_config to avoid a race condition with potential async update in the store
+        state.event_source_mappings[esm_config["UUID"]] = esm_config.copy()
+        function_version = LambdaProvider._get_function_version_from_arn(function_arn)
+        function_role = function_version.config.role
+        enabled = request.get("Enabled", True)
+        # TODO: check for potential async race condition update -> think about locking
+        esm_worker = EsmWorkerFactory(esm_config, function_role, enabled).get_esm_worker()
+        self.esm_workers[esm_worker.uuid] = esm_worker
+        # TODO: check StateTransitionReason, LastModified, LastProcessingResult (concurrent updates requires locking!)
+        esm_worker.create()
+        return esm_config
+
+    def create_event_source_mapping_v1(
+        self, context: RequestContext, request: CreateEventSourceMappingRequest
+    ) -> EventSourceMappingConfiguration:
+        fn_arn, function_name, state = self.validate_event_source_mapping(context, request)
```

Comment:
`create_event_source_mapping_v1` is basically the old implementation but I extracted some basic validations into `self.validate_event_source_mapping(context, request)` to get more tests passing. We are missing API-level test coverage for better validations (the API surface is quite big)